data-prepper-plugins/key-value-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java

```diff
@@ -179,79 +240,215 @@ private void validateKeySets(final Set<String> includeSet, final Set<String> exc
-                    key = key.replaceAll(keyValueProcessorConfig.getDeleteKeyRegex(), "");
-                }
-                key = keyValueProcessorConfig.getPrefix() + key;
+            recordEvent.put(keyValueProcessorConfig.getDestination(), processedMap);
+        }
 
-                if (terms.length == 2) {
-                    value = terms[1];
-                } else {
-                    LOG.debug("Unsuccessful match: '{}'", terms[0]);
-                    value = keyValueProcessorConfig.getNonMatchValue();
-                }
+        return records;
+    }
+
+    private ObjectNode recurse(String input, ObjectMapper mapper) {
+        Stack<Character> bracketStack = new Stack<Character>();
+        Map<Character, Character> bracketMap = initBracketMap();
+        int pairStart = 0;
 
-                if(value != null
-                        && value instanceof String
-                        && keyValueProcessorConfig.getDeleteValueRegex() != null
-                        && !Objects.equals(keyValueProcessorConfig.getDeleteValueRegex(), "")) {
-                    value = ((String)value).replaceAll(keyValueProcessorConfig.getDeleteValueRegex(), "");
+        ArrayList<String> pairs = new ArrayList<String>();
+        ObjectNode root = mapper.createObjectNode();
+
+        for (int i = 0; i < input.length(); i++) {
+            if (bracketMap.containsKey(input.charAt(i))) {
+                bracketStack.push(input.charAt(i));
+            }
+
+            if (bracketMap.containsValue(input.charAt(i)) && !bracketStack.isEmpty()) {
+                if (bracketMap.get(bracketStack.peek()) == input.charAt(i)) {
+                    bracketStack.pop();
                 }
+            }
+
+            if (bracketStack.isEmpty() && input.charAt(i) == fieldDelimiterPattern.toString().charAt(0)) {
+                String pair = input.substring(pairStart, i);
+                pairs.add(pair);
+                pairStart = i + 1;
+            }
+        }
+
+        pairs.add(input.substring(pairStart));
+
+        for (final String pair : pairs) {
+            int keyStart = 0;
+            int keyEnd = -1;
+            int valueStart = -1;
+            int valueEnd = -1;
+            String keyString = "";
+            String valueString;
 
-                if (keyValueProcessorConfig.getWhitespace().equals(whitespaceStrict)) {
-                    String[] whitespace_arr = trimWhitespace(key, value);
-                    key = whitespace_arr[0];
-                    value = whitespace_arr[1];
+            bracketStack.clear();
+
+            for (int i = 0; i < pair.length(); i++) {
+                if (bracketStack.isEmpty() && pair.charAt(i) == keyValueDelimiterPattern.toString().charAt(0)) {
+                    keyString = pair.substring(keyStart, i).stripTrailing();
+                    valueStart = i + 1;
+                    while(pair.charAt(valueStart) == ' ') {
```

Comment:
Assign the whitespace into a static final variable instead of using it as a magic string.